File: tests/functions/native_cast/native_cast.fs
module rec native_cast.Program

// Unsafe pointer casting.

let main _ =
  let mutNull: voidptr = __nativeCast (unativeint 0)
  let constNull: __voidconstptr = __nativeCast mutNull
  let constIntPtr: __constptr<int> = __nativeCast constNull
  let intPtr: nativeptr<int> = __nativeCast constIntPtr

  let address: nativeint = __nativeCast intPtr
  assert (int address = 0)
  0

File: tests/functions/native_cast/native_cast.c
#include "milone.h"

int main();

int main() {
    void* mutNull_ = ((void*)((uintptr_t)0));
    void const* constNull_ = ((void const*)mutNull_);
    int const* constIntPtr_ = ((int const*)constNull_);
    int* intPtr_ = ((int*)constIntPtr_);
    intptr_t address_ = ((intptr_t)intPtr_);
    milone_assert((((int)address_) == 0), 11, 2);
    return 0;
}
